### ISSUE TITLE
Referencing Theme Colors with CSS Custom Properties

### ISSUE THREAD

**Issue by fcisio** [system-ui/theme-ui#1205]
**Is your feature request related to a problem? Please describe.**

I've been running into many challenges regarding the usage of theme colors. It works great when using the properties that are already linked to the `colors` theme key, but this isn't always flexible enough.

For example, the variant of a complex component can get messy really quickly if the background-color need's to be in a pseudo-element or something like that.

![Shot_0 15](https://user-images.githubusercontent.com/46095852/94984156-26e72000-0517-11eb-80d2-d7052a31db44.jpg)

Also, the theme custom properties can't actually be used like `var(--theme-ui-colors-primary)`, which doesn't help.

**Describe the solution you'd like**

I'm thinking about linking theme keys to custom properties by checking if that property contains a color related key.
Example: --**`color`**-random-name || --**`bg`**-of-something || --**`fill`**-of-something || ...

```js
'--color-base': "primary",
'--color-hover': "text",
```
> This pattern could be implemented with many other properties too

This would allow for cleaner and more flexible variants.

![Shot_0 16](https://user-images.githubusercontent.com/46095852/94984356-26e81f80-0519-11eb-89a7-05779153ec73.jpg)

**Describe alternatives you've considered**
I initially wanted to just use the custom properties from the theme, but because of their shape, they can't be used outside of Theme UI.

Changing the shape of those variables would allow using the theme defined colors, by simply referencing the `var()`.

Should change from:
`--theme-ui-colors-primary: var(--theme-ui-colors-primary,#0759F9);`
to:
`--theme-ui-colors-primary: #0759F9;`

**Additional context**
Another use case that would benefit from this would be theming SVG's.
By running SVG's through SVGO I can replace set color codes with custom properties. Then I would create color sets for those, but can't access them on the SVG.

**Comment by fcisio** [comment]
Hi, there hasn't been any activity on this so I wanted to see if there have been any thoughts on this yet.

I feel like it could be fairly easy to implement and would have a huge payoff (at least the alternative solution).

> I have since then found an alternative using the theme context to define custom CSS properties, but it is still less than ideal.

Thanks in advance.

**Comment by atanasster** [comment]
Hi @fcisio - sorry, probably I haven't encountered the same use case and I am not sure I understand exactly your request.

For you first solution - can you post a theme - how you would like to write it, and how that theme should be actually transformed. Or if you prefer to start a PR, I will be glad to help if needed.

On the other hand, I also started a POC babel plugin that transforms theme-ui themes, based on string based custom keys: 
https://github.com/atanasster/babel-plugin-theme-ui
The plugin is currently not operational as a plugin, but you can use the test suites to try it out. Maybe its along the lines of what you needed?

**Comment by fcisio** [comment]
Hey @atanasster 
Thanks for the reply. I took a look at the Babel Plugin and I really like the idea.

This would indeed be a great solution for this use case since you would be able to reference theme values on any properties.
I think this is the most reliable approach.

**So I could do something like this:**
```js
export const theme = {
  colors: {
    primary: '#FFFFFF',
    secondary: '#000000',
  },
  input: {
   '--bg': 'colors.primary',
   '--bg-hover': 'colors.secondary'
  },
};
```
---

As for my suggested approach, here's an example of the source and output, just in case it can be helpful.

```js
export const theme = {
  colors: {
    primary: '#FFFFFF',
    secondary: '#000000',
  },
  input: {
   bg: 'primary',
   fontFamily: 'heading',
   '--bg-hover': 'secondary',
   '--color-hover': 'primary',
   '--bg-random': 'secondary',
   '--color': 'primary',
   '--fontFamily': 'heading',
   '--space-gap': [2, 3, 4],
  },
};
```

```js
export const theme = {
  colors: {
    primary: '#FFFFFF',
    secondary: '#000000',
  },
  input: {
   bg: '#FFFFFF',
   fontFamily: 'heveltica, arial, sans-serif',
   '--bg-hover': '#000000',
   '--color-hover': '#FFFFFF',
   '--bg-random': '#000000',
   '--color': '#FFFFFF',
   '--fontFamily': 'heveltica, arial, sans-serif',
   '--space-gap': ['15px', '30px', '45px'],
  },
};
```

Basically `color` and `--color` would both reference the theme colors. This could work with any CSS Custom Property containing the right property key.

- color / --color
- color / --color
- color / --color

**Comment by atanasster** [comment]
@fcisio - thanks a lot, I created a new thread #1234 where we can discuss the specs and work together on those transformations.

**Comment by atanasster** [comment]
guys, I think we can close this issue and consolidate all conversation to #1234 or directly to the babel plugin repo
